llvm-profdata.1
.\" Man page generated from reStructuredText.
.
.TH "LLVM-PROFDATA" "1" "2016-08-07" "clang-800" "LLVM"
.SH NAME
llvm-profdata \- Profile data tool
.
.nr rst2man-indent-level 0
.
.de1 rstReportMargin
\\$1 \\n[an-margin]
level \\n[rst2man-indent-level]
level margin: \\n[rst2man-indent\\n[rst2man-indent-level]]
-
\\n[rst2man-indent0]
\\n[rst2man-indent1]
\\n[rst2man-indent2]
..
.de1 INDENT
.\" .rstReportMargin pre:
. RS \\$1
. nr rst2man-indent\\n[rst2man-indent-level] \\n[an-margin]
. nr rst2man-indent-level +1
.\" .rstReportMargin post:
..
.de UNINDENT
. RE
.\" indent \\n[an-margin]
.\" old: \\n[rst2man-indent\\n[rst2man-indent-level]]
.nr rst2man-indent-level -1
.\" new: \\n[rst2man-indent\\n[rst2man-indent-level]]
.in \\n[rst2man-indent\\n[rst2man-indent-level]]u
..
.SH SYNOPSIS
.sp
\fBllvm\-profdata\fP \fIcommand\fP [\fIargs...\fP]
.SH DESCRIPTION
.sp
The \fBllvm\-profdata\fP tool is a small utility for working with profile
data files.
.SH COMMANDS
.INDENT 0.0
.IP \(bu 2
\fI\%merge\fP
.IP \(bu 2
\fI\%show\fP
.UNINDENT
.SH MERGE
.SS SYNOPSIS
.sp
\fBllvm\-profdata merge\fP [\fIoptions\fP] [\fIfilename...\fP]
.SS DESCRIPTION
.sp
\fBllvm\-profdata merge\fP takes several profile data files
generated by PGO instrumentation and merges them together into a single
indexed profile data file.
.sp
By default profile data is merged without modification. This means that the
relative importance of each input file is proportional to the number of samples
or counts it contains. In general, the input from a longer training run will be
interpreted as relatively more important than a shorter run. Depending on the
nature of the training runs it may be useful to adjust the weight given to each
input file by using the \fB\-weighted\-input\fP option.
.SS OPTIONS
.INDENT 0.0
.TP
.B \-help
Print a summary of command line options.
.UNINDENT
.INDENT 0.0
.TP
.B \-output=output, \-o=output
Specify the output file name.  \fIOutput\fP cannot be \fB\-\fP as the resulting
indexed profile data can\(aqt be written to standard output.
.UNINDENT
.INDENT 0.0
.TP
.B \-weighted\-input=weight,filename
Specify an input file name along with a weight. The profile counts of the input
file will be scaled (multiplied) by the supplied \fBweight\fP, where where \fBweight\fP
is a decimal integer >= 1. Input files specified without using this option are
assigned a default weight of 1. Examples are shown below.
.UNINDENT
.INDENT 0.0
.TP
.B \-instr (default)
Specify that the input profile is an instrumentation\-based profile.
.UNINDENT
.INDENT 0.0
.TP
.B \-sample
Specify that the input profile is a sample\-based profile.
.sp
The format of the generated file can be generated in one of three ways:
.INDENT 7.0
.TP
.B \-binary (default)
.UNINDENT
.sp
Emit the profile using a binary encoding. For instrumentation\-based profile
the output format is the indexed binary format.
.INDENT 7.0
.TP
.B \-text
.UNINDENT
.sp
Emit the profile in text mode. This option can also be used with both
sample\-based and instrumentation\-based profile. When this option is used
the profile will be dumped in the text format that is parsable by the profile
reader.
.INDENT 7.0
.TP
.B \-gcc
.UNINDENT
.sp
Emit the profile using GCC\(aqs gcov format (Not yet supported).
.INDENT 7.0
.TP
.B \-sparse[=true|false]
.UNINDENT
.sp
Do not emit function records with 0 execution count. Can only be used in
conjunction with \-instr. Defaults to false, since it can inhibit compiler
optimization during PGO.
.UNINDENT
.SS EXAMPLES
.SS Basic Usage
.sp
Merge three profiles:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
llvm\-profdata merge foo.profdata bar.profdata baz.profdata \-output merged.profdata
.ft P
.fi
.UNINDENT
.UNINDENT
.SS Weighted Input
.sp
The input file \fIfoo.profdata\fP is especially important, multiply its counts by 10:
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
llvm\-profdata merge \-weighted\-input=10,foo.profdata bar.profdata baz.profdata \-output merged.profdata
.ft P
.fi
.UNINDENT
.UNINDENT
.sp
Exactly equivalent to the previous invocation (explicit form; useful for programmatic invocation):
.INDENT 0.0
.INDENT 3.5
.sp
.nf
.ft C
llvm\-profdata merge \-weighted\-input=10,foo.profdata \-weighted\-input=1,bar.profdata \-weighted\-input=1,baz.profdata \-output merged.profdata
.ft P
.fi
.UNINDENT
.UNINDENT
.SH SHOW
.SS SYNOPSIS
.sp
\fBllvm\-profdata show\fP [\fIoptions\fP] [\fIfilename\fP]
.SS DESCRIPTION
.sp
\fBllvm\-profdata show\fP takes a profile data file and displays the
information about the profile counters for this file and
for any of the specified function(s).
.sp
If \fIfilename\fP is omitted or is \fB\-\fP, then \fBllvm\-profdata show\fP reads its
input from standard input.
.SS OPTIONS
.INDENT 0.0
.TP
.B \-all\-functions
Print details for every function.
.UNINDENT
.INDENT 0.0
.TP
.B \-counts
Print the counter values for the displayed functions.
.UNINDENT
.INDENT 0.0
.TP
.B \-function=string
Print details for a function if the function\(aqs name contains the given string.
.UNINDENT
.INDENT 0.0
.TP
.B \-help
Print a summary of command line options.
.UNINDENT
.INDENT 0.0
.TP
.B \-output=output, \-o=output
Specify the output file name.  If \fIoutput\fP is \fB\-\fP or it isn\(aqt specified,
then the output is sent to standard output.
.UNINDENT
.INDENT 0.0
.TP
.B \-instr (default)
Specify that the input profile is an instrumentation\-based profile.
.UNINDENT
.INDENT 0.0
.TP
.B \-text
Instruct the profile dumper to show profile counts in the text format of the
instrumentation\-based profile data representation. By default, the profile
information is dumped in a more human readable form (also in text) with
annotations.
.UNINDENT
.INDENT 0.0
.TP
.B \-sample
Specify that the input profile is a sample\-based profile.
.UNINDENT
.SH EXIT STATUS
.sp
\fBllvm\-profdata\fP returns 1 if the command is omitted or is invalid,
if it cannot read input files, or if there is a mismatch between their data.
.SH AUTHOR
Maintained by The LLVM Team (http://llvm.org/).
.SH COPYRIGHT
2003-2016, LLVM Project
.\" Generated by docutils manpage writer.
.


src/tools/llvm-profdata/llvm-profdata.cpp
//===- llvm-profdata.cpp - LLVM profile data tool -------------------------===//
//
//                     The LLVM Compiler Infrastructure
//
// This file is distributed under the University of Illinois Open Source
// License. See LICENSE.TXT for details.
//
//===----------------------------------------------------------------------===//
//
// llvm-profdata merges .profdata files.
//
//===----------------------------------------------------------------------===//

#include "llvm/ADT/OwningPtr.h"
#include "llvm/ADT/StringRef.h"
#include "llvm/ProfileData/InstrProfReader.h"
#include "llvm/ProfileData/InstrProfWriter.h"
#include "llvm/Support/CommandLine.h"
#include "llvm/Support/Format.h"
#include "llvm/Support/ManagedStatic.h"
#include "llvm/Support/MemoryBuffer.h"
#include "llvm/Support/PrettyStackTrace.h"
#include "llvm/Support/Signals.h"
#include "llvm/Support/raw_ostream.h"

using namespace llvm;

static void exitWithError(const Twine &Message, StringRef Whence = "") {
  errs() << "error: ";
  if (!Whence.empty())
    errs() << Whence << ": ";
  errs() << Message << "\n";
  ::exit(1);
}

int merge_main(int argc, const char *argv[]) {
  cl::list<std::string> Inputs(cl::Positional, cl::Required, cl::OneOrMore,
                               cl::desc("<filenames...>"));

  cl::opt<std::string> OutputFilename("output", cl::value_desc("output"),
                                      cl::init("-"),
                                      cl::desc("Output file"));
  cl::alias OutputFilenameA("o", cl::desc("Alias for --output"), cl::Required,
                            cl::aliasopt(OutputFilename));

  cl::ParseCommandLineOptions(argc, argv, "LLVM profile data merger\n");

  if (OutputFilename.compare("-") == 0)
    exitWithError("Cannot write indexed profdata format to stdout.");

  std::string ErrorInfo;
  raw_fd_ostream Output(OutputFilename.data(), ErrorInfo, sys::fs::F_None);
  if (!ErrorInfo.empty())
    exitWithError(ErrorInfo, OutputFilename);

  InstrProfWriter Writer;
  for (const auto &Filename : Inputs) {
    std::unique_ptr<InstrProfReader> Reader;
    if (error_code ec = InstrProfReader::create(Filename, Reader))
      exitWithError(ec.message(), Filename);

    for (const auto &I : *Reader)
      if (error_code EC = Writer.addFunctionCounts(I.Name, I.Hash, I.Counts))
        errs() << Filename << ": " << I.Name << ": " << EC.message() << "\n";
    if (Reader->hasError())
      exitWithError(Reader->getError().message(), Filename);
  }
  Writer.write(Output);

  return 0;
}

int show_main(int argc, const char *argv[]) {
  cl::opt<std::string> Filename(cl::Positional, cl::Required,
                                cl::desc("<profdata-file>"));

  cl::opt<bool> ShowCounts("counts", cl::init(false),
                           cl::desc("Show counter values for shown functions"));
  cl::opt<bool> ShowAllFunctions("all-functions", cl::init(false),
                                 cl::desc("Details for every function"));
  cl::opt<std::string> ShowFunction("function",
                                    cl::desc("Details for matching functions"));

  cl::opt<std::string> OutputFilename("output", cl::value_desc("output"),
                                      cl::init("-"),
                                      cl::desc("Output file"));
  cl::alias OutputFilenameA("o", cl::desc("Alias for --output"),
                            cl::aliasopt(OutputFilename));

  cl::ParseCommandLineOptions(argc, argv, "LLVM profile data summary\n");

  std::unique_ptr<InstrProfReader> Reader;
  if (error_code EC = InstrProfReader::create(Filename, Reader))
    exitWithError(EC.message(), Filename);

  if (OutputFilename.empty())
    OutputFilename = "-";

  std::string ErrorInfo;
  raw_fd_ostream OS(OutputFilename.data(), ErrorInfo);
  if (!ErrorInfo.empty())
    exitWithError(ErrorInfo, OutputFilename);

  if (ShowAllFunctions && !ShowFunction.empty())
    errs() << "warning: -function argument ignored: showing all functions\n";

  uint64_t MaxFunctionCount = 0, MaxBlockCount = 0;
  size_t ShownFunctions = 0, TotalFunctions = 0;
  for (const auto &Func : *Reader) {
    bool Show = ShowAllFunctions ||
                (!ShowFunction.empty() &&
                 Func.Name.find(ShowFunction) != Func.Name.npos);

    ++TotalFunctions;
    assert(Func.Counts.size() > 0 && "function missing entry counter");
    if (Func.Counts[0] > MaxFunctionCount)
      MaxFunctionCount = Func.Counts[0];

    if (Show) {
      if (!ShownFunctions)
        OS << "Counters:\n";
      ++ShownFunctions;

      OS << "  " << Func.Name << ":\n"
         << "    Hash: " << format("0x%016" PRIx64, Func.Hash) << "\n"
         << "    Counters: " << Func.Counts.size() << "\n"
         << "    Function count: " << Func.Counts[0] << "\n";
    }

    if (Show && ShowCounts)
      OS << "    Block counts: [";
    for (size_t I = 1, E = Func.Counts.size(); I < E; ++I) {
      if (Func.Counts[I] > MaxBlockCount)
        MaxBlockCount = Func.Counts[I];
      if (Show && ShowCounts)
        OS << (I == 1 ? "" : ", ") << Func.Counts[I];
    }
    if (Show && ShowCounts)
      OS << "]\n";
  }
  if (Reader->hasError())
    exitWithError(Reader->getError().message(), Filename);

  if (ShowAllFunctions || !ShowFunction.empty())
    OS << "Functions shown: " << ShownFunctions << "\n";
  OS << "Total functions: " << TotalFunctions << "\n";
  OS << "Maximum function count: " << MaxFunctionCount << "\n";
  OS << "Maximum internal block count: " << MaxBlockCount << "\n";
  return 0;
}

int main(int argc, const char *argv[]) {
  // Print a stack trace if we signal out.
  sys::PrintStackTraceOnErrorSignal();
  PrettyStackTraceProgram X(argc, argv);
  llvm_shutdown_obj Y; // Call llvm_shutdown() on exit.

  StringRef ProgName(sys::path::filename(argv[0]));
  if (argc > 1) {
    int (*func)(int, const char *[]) = 0;

    if (strcmp(argv[1], "merge") == 0)
      func = merge_main;
    else if (strcmp(argv[1], "show") == 0)
      func = show_main;

    if (func) {
      std::string Invocation(ProgName.str() + " " + argv[1]);
      argv[1] = Invocation.c_str();
      return func(argc - 1, argv + 1);
    }

    if (strcmp(argv[1], "-h") == 0 ||
        strcmp(argv[1], "-help") == 0 ||
        strcmp(argv[1], "--help") == 0) {

      errs() << "OVERVIEW: LLVM profile data tools\n\n"
             << "USAGE: " << ProgName << " <command> [args...]\n"
             << "USAGE: " << ProgName << " <command> -help\n\n"
             << "Available commands: merge, show\n";
      return 0;
    }
  }

  if (argc < 2)
    errs() << ProgName << ": No command specified!\n";
  else
    errs() << ProgName << ": Unknown command!\n";

  errs() << "USAGE: " << ProgName << " <merge|show> [args...]\n";
  return 1;
}
